File: sdk/platform/Web/src/QRDetection.cpp
#include <Posemesh/QRDetection.hpp>
#include <emscripten/bind.h>
#include <memory>

using namespace emscripten;
using namespace psm;

namespace {
bool detectQRFromLuminance(
    const std::vector<uint8_t>& imageBytes,
    int width,
    int height,
    std::vector<std::string>& outContents,
    std::vector<std::shared_ptr<Vector2>>& outCorners)
{
    std::vector<Vector2> corners;
    bool detectionResult = QRDetection::detectQRFromLuminance(imageBytes, width, height, outContents, corners);

    outCorners.clear();
    for (auto& corner : corners) {
        outCorners.emplace_back(new Vector2(std::move(corner)));
    }

    return detectionResult;
}

std::vector<std::shared_ptr<psm::LandmarkObservation>> detectQRFromLuminanceLandmarkObservations(
    const std::vector<uint8_t>& imageBytes,
    int width,
    int height)
{
    std::vector<std::shared_ptr<psm::LandmarkObservation>> observations;
    auto lo = QRDetection::detectQRFromLuminance(imageBytes, width, height);    
    for (auto& observation : lo) {
        observations.emplace_back(new psm::LandmarkObservation(std::move(observation)));
    }

    return observations;
}
}

EMSCRIPTEN_BINDINGS(QRDetection)
{
    class_<QRDetection>("QRDetection")
        .class_function("__detectQRFromLuminance(imageBytes, width, height, outContents, outCorners)", &detectQRFromLuminance)
        .class_function("__detectQRFromLuminanceLandmarkObservations(imageBytes, width, height)", &detectQRFromLuminanceLandmarkObservations);
}

File: sdk/platform/Web/QRDetection.js
posemeshModule.QRDetection = null;

__internalPosemeshAPI.builderFunctions.push(function() {
    __internalPosemesh.QRDetection.detectQRFromLuminance = function(imageBytes, width, height, outContents, outCorners) {
        let imageBytesVector = undefined, outContentsVector = undefined, outCornersVector = undefined;
        try {
            imageBytesVector = __internalPosemeshAPI.toVectorUint8(imageBytes);
            outContentsVector = new __internalPosemesh.VectorString();
            outCornersVector = new __internalPosemesh.VectorVector2();
            let result = __internalPosemesh.QRDetection.__detectQRFromLuminance(imageBytesVector, width, height, outContentsVector, outCornersVector);

            if (result) {
                outContents.length = 0;
                let outContentsVectorSize = outContentsVector.size();
                for (let i = 0; i < outContentsVectorSize; i++) {
                    outContents.push(outContentsVector.get(i));
                }

                for (let corner of outCorners) {
                    corner.delete();
                }
                outCorners.length = 0;
                let outCornersVectorSize = outCornersVector.size();
                for (let i = 0; i < outCornersVectorSize; i++) {
                    outCorners.push(outCornersVector.get(i));
                }
            }

            return result;
        } finally {
            if (outCornersVector) {
                outCornersVector.delete();
            }
            if (outContentsVector) {
                outContentsVector.delete();
            }
            if (imageBytesVector) {
                imageBytesVector.delete();
            }
        }
    };

    __internalPosemesh.QRDetection.detectQRFromLuminanceLandmarkObservations = function(imageBytes, width, height) {
        let imageBytesVector = undefined;
        try {
            imageBytesVector = __internalPosemeshAPI.toVectorUint8(imageBytes)
            let result = []
            let observations = __internalPosemesh.QRDetection.__detectQRFromLuminanceLandmarkObservations(imageBytesVector, width, height)

            for (let i = 0; i < observations.size(); i++) {
                result.push(observations.get(i))
            }

            return result;
        } finally {
            if (imageBytesVector) {
                imageBytesVector.delete();
            }
        }
    };

    posemeshModule.QRDetection = __internalPosemesh.QRDetection;
});
